### PR TITLE
[3951] Upgrade to Rails 7

### PR DESCRIPTION
**Pull request by jhenderson**
### Context
https://trello.com/c/EcZJLgSF/3951-l-upgrade-to-rails-7023

### Changes proposed in this pull request
- Updated the `rails` gem to `7.0.4`
- Moved course and subject constants into separate folder under `config/subjects/`
- Various fixes to get all specs passing

### Guidance to review
- Run the app locally and play around with the UI to double check everything is good.

### Important business

- [ ] Does this PR introduce any PII fields that need to be overwritten or deleted in db/scripts/sanitise.sql?
- [ ] Does this PR change the database schema? If so, have you updated the config/analytics.yml file and considered whether you need to send 'import_entity' events?

NB: Please notify the #twd_data_insights team and ask for a review if new fields are being added to analytics.yml

### CODE REVIEW

File: app/services/degrees/dfe_reference.rb [hunk 1]
```diff
@@ -1,7 +1,7 @@
 # frozen_string_literal: true
 
 module Degrees
-  class DfeReference
+  class DfEReference
```
Comment (jhenderson) [inline app/services/degrees/dfe_reference.rb:4]: Rails 7 forced the rename here.

File: config/initializers/subjects/course_subjects.rb [hunk 2]
```diff
@@ -0,0 +1,162 @@
+# frozen_string_literal: true
+
+require "degree_subjects"
+
+unless defined?(CourseSubjects)
```
Comment (jhenderson) [inline config/initializers/subjects/course_subjects.rb:5]: Not ideal, but it allowed me to get around issues rails 7 had loading these constants in other directories such as `lib/` and not doing it in a way that raises warnings about constants are loaded.

File: config/application.rb [hunk 3]
```diff
@@ -25,8 +25,7 @@
 
 module RegisterTraineeTeachers
   class Application < Rails::Application
-    # Initialize configuration defaults for originally generated Rails version.
-    config.load_defaults(5.2)
+    config.load_defaults(7.0)
```
Comment (jhenderson) [inline config/application.rb:28]: If we use the `7.0` defafaults, then we should be able to remove the following files:

```
new_framework_defaults_5_2.rb
new_framework_defaults_6_0.rb
new_framework_defaults_7_0.rb
```

Any objections?

File: app/controllers/personas_controller.rb [hunk 4]
```diff
@@ -1,10 +1,9 @@
 # frozen_string_literal: true
 
-unless FeatureService.enabled?("use_dfe_sign_in")
-  class PersonasController < ApplicationController
-    skip_before_action :authenticate
-    def index
-      @personas = Persona.all.order(:first_name)
-    end
+class PersonasController < ApplicationController
```
Comment (jhenderson) [inline app/controllers/personas_controller.rb:3]: Rails 7 didn't like the controller  wrapped in a guard clause. Better this way.

File: Gemfile [hunk 5]
```diff
@@ -162,9 +163,8 @@ group :development do
   gem "web-console", ">= 3.3.0"
 
   # Spring speeds up development by keeping your application running in the background. Read more: https://github.com/rails/spring
-  gem "spring"
+  gem "spring", "~> 4.0"
   gem "spring-commands-rspec", "~> 1.0"
-  gem "spring-watcher-listen", "~> 2.0.0"
```
Comment (jhenderson) [inline Gemfile:167]: Had to remove this gem if we want to upgrade to `spring ~>4`. Any objects? Is anyone using it?

**Comment by jhenderson** [comment]
[[L] Upgrade to Rails 7.0.2.3](https://trello.com/c/EcZJLgSF/3951-l-upgrade-to-rails-7023)

**Review by AbigailMcP — COMMENTED** [review]
Tested it out on the review app and ran through all the basic functionality, csv exports etc. 
Everything looks good except the JS! The filters seem to broken :(

File: app/controllers/base_trainee_controller.rb [hunk 6]
```diff
@@ -227,6 +227,6 @@ def json_response
   end
 
   def render_partial(partial, locals)
-    (render_to_string(formats: %w[html], partial: partial, locals: locals) || "").squish
+    (render_to_string(formats: [:html], partial: partial, locals: locals) || "").squish
```
Comment (jhenderson) [inline app/controllers/base_trainee_controller.rb:230]: @AbigailMcP This was the culprit preventing the filters from working. Rails 7 apparently only accepts symbols.

**Comment by jhenderson** [comment]
> Tested it out on the review app and ran through all the basic functionality, csv exports etc. Everything looks good except the JS! The filters seem to broken :(

Good spot @AbigailMcP - fixed now, see my comment above.

**Review by AbigailMcP — APPROVED** [review]
Phew, I'm glad the filter bug wasn't too tricky! Nice job 🚢

**Review by edwardhorsford — APPROVED** [review]

**Pull request merged**